### ISSUE TITLE
Roc's Cape: Add momentum

### ISSUE THREAD

**Issue by MetalES** [MetalES/Project-Zelda#51]
In Minish Cap, Roc's cape has momentum, if you release the direction input, you continue in the direction you pressed. Which get me to this point.

on_draw's refresh rate is lower then on_update (each frame vs each cycle), so why not using it ?

We need:

- [ ] Add a variable for X and Y momentum
- [ ] Reset these variable when landed or when down thrusting
- [ ] Use hero:set_position and apply get_position + momentum on the hero
- [ ]  Hero:test_obstacles(2, 2)

Just test this, just to see how it looks in game

**Comment by MetalES** [comment]
Momentum is never kept, even if I tell the variable to memorize this value + decrease over time, as soon as the input is released, the momentum is canceled

**Comment by MetalES** [comment]
This should looks like this, now it is time to do some conditions

`function roc_cape:manage_momentum()
  local cx, cy = 8, 8
  local hero = self.hero
  local direction = hero:get_direction()
  
  self.momentum_x = 0
  self.momentum_y = 0
  
  -- Give a little boost if the input is pressed
  if game:is_command_pressed(self.direction[direction]) then
	if direction % 2 == 0 then
	  self.momentum_x = 0.5
	else 
	  self.momentum_y = 0.5
	end
  end
  
  sol.timer.start(self, 10, function()
    local x, y = hero:get_position()
	-- Do everything here
	
	x = x + self.momentum_x
	y = y + self.momentum_y
	hero:set_position(x, y)
	return true
  end)
  
end
`

**Comment by Diarandor** [comment]
I haven't tested this but, try to use a stream that follows the hero. It may work to simulate the inertia.

**Comment by MetalES** [comment]
This would be difficult, the hero can go backward and speed isn't constant, but I take this idea in note, this is clever indeed.

**Comment by Diarandor** [comment]
I don't know that either. But still, if you are using a custom entity with traversable ground under the hero during the jump, and the stream is created later, it should be above the traversable ground instead of over the hole. (Note that you can also jump over lava and deep water. I guess the streams should work over lava, but who knows... XD)

**Comment by Diarandor** [comment]
Another idea (and quite easy) for the speed. Just use a timer to check the direction of the hero, or the direction of the movement, each few milliseconds, and set a slow speed when the direction is not the initial one. That may work too. But if you want to add diagonal jumps, which is convenient, then you should check the direction of the initial movement instead of the direction of the hero. I hope this helps.

**Comment by MetalES** [comment]
I already use the second idea, but you have to move constantly the hero by a movement, the stream idea is clever and it actually works, now I should try to disable Link's movement in order to make the jump handled by the stream.
Good thig is streams allow diagonals and there is a game event that catch if we are moving diagonal (game:get_commands_direction())

**Comment by Diarandor** [comment]
I was wondering if having several streams under the hero may be a problem if only one of them is applied, because that would interfere with the behavior of the fan stream. The second idea (changing the speed) is probably the best idea, and should work perfectly; in this case I would set a speed that depends (in a smoothly way) on the angle of the initial direction of the movement of the hero and the current direction of the movement of the hero, i.e., the bigger angle difference, the less speed (and maybe with a minimum speed).

**Comment by MetalES** [comment]
Yeah overlapping another stream has a weird issue, I should try a movement with a speed that is constantly repeted until the item is finished

**Comment by Diarandor** [comment]
Are you freezing and disabling the hero's movement? That sounds bad. Didn't you try to modify the speed of the movement (depending on the angle) as I suggested?

**Comment by MetalES** [comment]
Using a straight movement there are weird stuffs happening, sometimes it works, sometimes not, I am not freezing the hero for this method (i was freezing him with the stream method)

**Comment by Diarandor** [comment]
Ok, so if I understand correctly, you were forcing a movement during the jump. The problem of that is that you cannot control the movement in the air since the hero would be frozen, unless you use "set_position" instead of a built-in movement (as you say above), which should be a nice solution.

I don't understand what you mean by "to set a fixed animation you have to unfreeze the hero".

**Comment by MetalES** [comment]
In your script, if I want to use a fixed animation through hero:set_fixed_animation(), I need to unfreeze him to display the right animation. This might be normal.

Changing the speed of the movement works but with minor bugs, sometimes it work, sometimes not.

What I want to do is:

- Create 2 variables that get the X and Y inertia (the value added to the hero's x and y in set_position()), which are increased or decreased, depending  if the player press an input.
- Decrease the X and Y value when the input is released

This seems simple, it would looks like how it looks in Minish Cap, technically speaking, movement will be managed 

This is how the code looks right now, the only missing thing is input checking to increment / decrement the inertia, this might be it

`function roc_cape:manage_inertia()
  local game = self.game
  local hero = self.hero
  local cx, cy = 0, 0
  
  local right = game:is_command_pressed("right")
  local up = game:is_command_pressed("up")
  local left = game:is_command_pressed("left")
  local down = game:is_command_pressed("down")
  
  if right then 
    cx = 8
  elseif left then 
    cx = -8
  elseif up then 
    cy = -8
  elseif down then 
    cy = 8
  end
  
  -- Check what is in front of Link
  if hero:test_obstables(cx, cy) then
    -- Colliding up / down: cancel Y inertia
    if up or down then
	  self.inertia_y = 0
	end
   
     -- Colliding left / right: cancel X inertia
    if left or right then
	  self.inertia_x = 0
	end
  end

  -- Apply inertia depending on the input pressed AND the hero direction
  local direction = self.hero_direction
  local direction_input = self.direction[direction]
  
  
  -- Get Link position and apply inertia as well
  local x, y = hero:get_position()
  x = x + self.inertia_x
  y = y + self.inertia_y
  
  -- Set Link position
  -- Don't use layer, there is a special function to interact with upper layer in start_jump()
  hero:set_position(x, y)`

**Comment by MetalES** [comment]
This is what I am trying to be close of, example using images
This is how a jump looks like using the Roc's Cape currently.

We are pressing right then down, as shown bellow.
![sans titre 10](https://cloud.githubusercontent.com/assets/4288633/21557498/3a763c66-ce2d-11e6-823f-286d28307884.png)

The problem is, modifying the hero's walking speed is problematic, if he is stopped then the jump movement is as well, an inertia using a straight movement only imply the direction the script memorized when using the item.

This is how I see the Roc's cape with inertia (default idea)

The solution is, like Minish Cap, to improve maniability, this.
![sans titre 11](https://cloud.githubusercontent.com/assets/4288633/21557526/a5f5404a-ce2d-11e6-9211-c55ce2f45590.png)

This is the same movement, but this time, using 2 values, let's call them InertiaX and InertiaY, constantly updated depending on the input pressed and Link's direction
This looks better on the paper doesn't it, it would make the jump more realistic

**Comment by MetalES** [comment]
Yet I don't understand why a stream overlapping another causes malfunction, it was the best solution out there and it worked quite great, I' gonna stuck with the stream solution, the simplest, because fan streams aren't disabled when their speed = 0.
Or else, replace streams by a custom entity where, when overlapping it, set the hero's position to the right coordinate, a little bit like your Platform script, diarandor

**Comment by Diarandor** [comment]
Awesome work!!!
Another idea: Appart from modulating the stream speed of the fan (to make a smoother transition) and adding a semitranspartent wind effect, it would be nice also to make the sound transition smooth (when  the wind is starting or finishing, the wind sound could be lower); that would be even more realistic.

**Comment by MetalES** [comment]
The sound and sprite are planned, the sound is already implemented (See Roc Cape Test video), for the sprite I could use a code like the Hookshot's chain in order to make the wind animation progressive (https://github.com/MetalES/Project-Zelda/blob/master/data/scripts/gameplay/hero/hookshot_controller.lua#L340)

For the speed modulation do you have any idea ? The movement is managed in add_collision_test and the speed is managed through entity.speed = speed

**Comment by Diarandor** [comment]
Ok, I didn't notice, nice to know that. I can't help more for now, sorry; I must focus in the creation of new free graphics, it's my main priority (maybe one day in a few years you can release your game with free graphics, and new story/enemies, that would be great).

**Comment by MetalES** [comment]
The last thing I need is speed modulator, I don't know how to do it, everything is handled in add_collision_test, do you have any idea ?

**Comment by MetalES** [comment]
Oh wait wait wait, a fan in the project is composed of 4 states (disabled, start_enable, enabled and start_disable), perhaps I could play with a timer in start_enable/disable and disable / enable collision ... in each recheck
